### PR TITLE
WIP - reverse iterators

### PR DESCRIPTION
**Pull request by jlandrews**
Seems to work, not sure if there's a better way to seek to last key for goleveldb.

Also needs more testing to check edge cases.

### CODE REVIEW

**Comment by xla** [comment]
@jlandrews This repo is going to collapse into tendermint proper. Which should be done by tomorrow, can you resubmit your PR there?

File: db/c_level_db.go [hunk 1]
```diff
@@ -204,7 +205,11 @@ type cLevelDBIterator struct {
 
 func newCLevelDBIterator(source *levigo.Iterator, start, end []byte, isReverse bool) *cLevelDBIterator {
 	if isReverse {
-		panic("not implemented yet") // XXX
+		if end != nil {
```
Comment (jaekwon) [inline db/c_level_db.go]: tmlibs/db/types.go docs: 
```go
    // Iterate over a domain of keys in descending order. End is exclusive.
    // Start must be greater than end, or the Iterator is invalid.
    // If start is nil, iterates from the last/greatest item (inclusive).
    // If end is nil, iterates up to the first/least item (inclusive).
    // CONTRACT: No writes may happen within a domain while an iterator exists over it.
    // CONTRACT: start, end readonly []byte
    ReverseIterator(start, end []byte) Iterator
```
So we want:

```go
if start == nil {
    if isReverse {
        source.SeekToLast()
    } else {
        source.SeektoFirst()
    }
} else {
    source.Seek(start)
}
```

File: db/c_level_db.go [hunk 2]
```diff
@@ -243,7 +248,12 @@ func (itr cLevelDBIterator) Valid() bool {
 	// If key is end or past it, invalid.
 	var end = itr.end
 	var key = itr.source.Key()
-	if end != nil && bytes.Compare(end, key) <= 0 {
+	if !itr.isReverse && end != nil && bytes.Compare(end, key) <= 0 {
+		itr.isInvalid = true
+		return false
+	}
+
+	if itr.isReverse && start != nil && bytes.Compare(start, key) > 0 {
```
Comment (jaekwon) [inline db/c_level_db.go]: I would opt for if {} else {} for clarity regarding the isReverse check.

**Comment by jlandrews** [comment]
Comments addressed. Still plan to add more extensive testing to catch edge cases.

File: db/go_level_db.go [hunk 3]
```diff
@@ -207,10 +208,13 @@ type goLevelDBIterator struct {
 var _ Iterator = (*goLevelDBIterator)(nil)
 
 func newGoLevelDBIterator(source iterator.Iterator, start, end []byte, isReverse bool) *goLevelDBIterator {
-	if isReverse {
-		panic("not implemented yet") // XXX
+	if isReverse && start == nil {
+		for !source.Last() {
```
Comment (jaekwon) [inline db/go_level_db.go]: This doesn't seem right.... I'll submit a fix to this branch.  https://github.com/syndtr/goleveldb/blob/master/leveldb/iterator/iter.go#L29

Comment (jlandrews) [reply db/go_level_db.go]: Don't know why I thought Last was just checking and not seeking... Thanks for catching that.
